engine/src/engine.ts

```diff
@@ -0,0 +1,181 @@
+import Board from './board'
+import Player from './player'
+
+interface EngineConfig {
+  boardSize: {
+    rows: number
+    columns: number
+  }
+  winLength: number
+  players: Player[]
+  firstPlayer: Player
+}
+
+export default class Engine {
+  private board: Board
+  private turnPlayer: Player
+  private config: EngineConfig
+
+  constructor(config: EngineConfig) {
+    if (config.players.length !== 2) {
+      throw new Error('Invalid number of players')
+    } else if (config.players[0].value === config.players[1].value) {
+      throw new Error('Players should have different values')
+    } else if (
+      !config.players.find((player) => player === config.firstPlayer)
+    ) {
+      throw new Error('First player should be one of the players')
+    }
+
+    this.board = new Board()
+    this.turnPlayer = config.firstPlayer
+    this.config = config
+  }
+
+  private init() {
+    const { boardSize } = this.config
+    this.board.init(boardSize.rows, boardSize.columns)
+  }
+
+  public start() {
+    this.init()
+  }
+
+  public getTurnPlayer() {
+    return this.turnPlayer
+  }
+
+  public play(player: Player, row: number, column: number) {
+    // check if the player is valid
+    if (!this.config.players.find((p) => p === player)) {
+      throw new Error('Invalid player')
+    }
+
+    // check if it's the player's turn
+    if (player !== this.turnPlayer) {
+      throw new Error('Not the turn player')
+    }
+
+    // check if the location is valid
+    if (row < 0 || row >= this.config.boardSize.rows) {
+      throw new Error('Invalid row')
+    }
+    if (column < 0 || column >= this.config.boardSize.columns) {
+      throw new Error('Invalid column')
+    }
+    if (this.board.getState()[row][column] !== 0) {
+      throw new Error('Location is already taken')
+    }
+
+    // update the board
+    this.board.update(row, column, player.value)
+
+    // update the turn player
+    const nextPlayerIndex = (this.config.players.indexOf(player) + 1) % 2
```

Comment:
I know that you are currently building for only 2 players.  Not sure how difficult it would be to generalize to n > 2, but maybe the # of players limit n should be in config.  Then when you write code that depends on n == 2, add assert(n == 2) which will document the code which you would need to change if you generalize.

Reply:
I don't think I will ever generalize to have n != 2 as the game is supposed to be a 2-player game.
However, I updated the code use `this.config.players.length` instead of the magic number `2`.